src/wix/WixToolset.Core/Link/AddDefaultSymbolsCommand.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.
+
+namespace WixToolset.Core.Link
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using WixToolset.Data;
+    using WixToolset.Data.Symbols;
+
+    internal class AddDefaultSymbolsCommand
+    {
+        public static readonly string WixStandardInstallFolder = "INSTALLFOLDER";
+        public static readonly string WixStandardInstallFolderParent = "ProgramFiles6432Folder";
+        public static readonly string WixStandardInstallFolderReference = "Directory:INSTALLFOLDER";
+
+        public AddDefaultSymbolsCommand(FindEntrySectionAndLoadSymbolsCommand find, List<IntermediateSection> sections)
+        {
+            this.Find = find;
+            this.Sections = sections;
+        }
+
+        public IntermediateSection EntrySection { get; }
+
+        public IEnumerable<IntermediateSection> Sections { get; }
+
+        public FindEntrySectionAndLoadSymbolsCommand Find { get; }
+
+        public void Execute()
+        {
+            if (this.Find.EntrySection.Type != SectionType.Package)
+            {
+                // Only packages...for now.
+                return;
+            }
+
+            var symbols = this.Sections.SelectMany(s => s.Symbols);
+            var referenceSymbols = symbols.OfType<WixSimpleReferenceSymbol>();
+            var directorySymbols = symbols.OfType<DirectorySymbol>();
+
+            if (referenceSymbols.Any(s => s.SymbolicName == WixStandardInstallFolderReference)
+                && !directorySymbols.Any(d => d.Id.Id == WixStandardInstallFolder))
+            {
+                // If there are any INSTALLFOLDER references, add a default one, using the
+                // first reference as the "canonical" reference for source line numbers.
+                this.AddSymbol(new DirectorySymbol(null, new Identifier(AccessModifier.Global, WixStandardInstallFolder))
+                {
+                    ParentDirectoryRef = WixStandardInstallFolderParent,
+                    Name = "!(bind.Property.Manufacturer) !(bind.Property.ProductName)",
+                    SourceName = ".",
+                });
+
+                this.AddSymbol(new WixSimpleReferenceSymbol(null, new Identifier(AccessModifier.Global, WixStandardInstallFolder))
+                {
+                    Table = "Directory",
+                    PrimaryKeys = WixStandardInstallFolderParent,
+                });
+            }
+
+            var upgradeSymbols = symbols.OfType<UpgradeSymbol>();
+            if (!upgradeSymbols.Any())
+            {
+                var packageSymbol = symbols.OfType<WixPackageSymbol>().SingleOrDefault();
+
+                if (packageSymbol?.UpgradeStrategy == WixPackageUpgradeStrategy.MajorUpgrade
+                    && !String.IsNullOrEmpty(packageSymbol?.UpgradeCode))
```

Comment:
I wonder if we should always add the default major upgrade symbols if the strategy is `majorUpgrade`? I suppose the issue is that `majorUpgrade` strategy is default in v5 and people probably have `MajorUpgrade` elements from v4 which would create a collision. :hurtrealbad:

Reply:
`if the strategy is majorUpgrade` and there are no UpgradeSymbols so an existing MajorUpgrade won't trigger this.